### DISCUSSION TITLE
Mapperly - must have an implementation part because it has accessibility modifiers.

### DISCUSSION THREAD

**Discussion by reddyabhijith8888** [riok/mapperly#1431] [Q&A]
I have a .net framework 4.6.1 Application, and im using few c# libraries which are targeting .net standard 2.0, Im using Mapperly to map my domain to External DTOS. 

**I am facing an issue when im trying to run my code in a docker container(windows), which im not able to get any build errors in my local Visual studio.**

My Sample Implementation of this is 

```csharp
[Mapper]
 public partial class CustomerMapper
 {
     public partial CustomerItem MapBLLRequestToDomain(Dealer request);
     public partial Dealer MapDomainRequestToBLL(CustomerItem request);
     public partial List<CustomerItem> MapBLLRequestToDomain (List<Dealer> request);
     public partial List<Dealer> MapDomainRequestToBLL (List<CustomerItem> request);
 }
 ```

this mapper is implemented in one library and is used in multiple libraries in the application, now when i build this i get the error - 
```
2024-07-30T20:01:16.7795588Z   D:\a\1\s\myproject\test\test\Mappers\CustomerMapper.cs(10,37): error CS8795: Partial method 'CustomerMapper.MapBLLRequestToDomain(Dealer)' must have an implementation part because it has accessibility modifiers. [D:\a\1\s\myproject\test\test.csproj]
2024-07-30T20:01:16.7796447Z   D:\a\1\s\myproject\test\test\Mappers\CustomerMapper.cs(12,43): error CS8795: Partial method 'CustomerMapper.MapBLLRequestToDomain(List<Dealer>)' must have an implementation part because it has accessibility modifiers. [D:\a\1\s\myproject\test\test.csproj]
2024-07-30T20:01:16.7797320Z   DD:\a\1\s\myproject\test\test\Mappers\CustomerMapper.cs(11,62): error CS8795: Partial method 'CustomerMapper.MapDomainRequestToBLL(CustomerItem)' must have an implementation part because it has accessibility modifiers. [D:\a\1\s\myproject\test\test.csproj]
2024-07-30T20:01:16.7798541Z   D:\a\1\s\myproject\test\test\Mappers\CustomerMapper.cs(13,68): error CS8795: Partial method 'CustomerMapper.MapDomainRequestToBLL(List<CustomerItem>)' must have an implementation part because it has accessibility modifiers. [D:\a\1\s\myproject\test\test.csproj]
```

**Comment by sveinhelge** [comment]
It might be these two lines. How can Mapperly differ these methods. The have the same signature. So there is no way for Mapperly to know whom to choose.

```csharp
public partial List MapBLLRequestToDomain(List request);
public partial List MapDomainRequestToBLL(List request);
```

**Reply by reddyabhijith8888** [reply]
hey @sveinhelge  im not sure why the above is not the same as my code, when i click on edit i can see the right methods,. but its wierdly showing something like that after update
![image](https://github.com/user-attachments/assets/6945953e-0223-4988-856a-ff9366e905b3)

**Reply by sveinhelge** [reply]
I belive this is related to the fact you have both list and single mapping. When you create List<T> mapping single is also included. The single method becomes a private method.

So if I uncomment the line _//public partial Person ToPerson(PersonDto source);_ I will get
Partial method 'Person ToPerson(PersonDto)' must have an implementation part because it has accessibility modifiers

```csharp
[Mapper]
public partial class PersonMapper
{
    //public partial Person ToPerson(PersonDto source);
    public partial List<Person> ToPerson(List<PersonDto> source);
}

public class PersonDto
{
    public int Id { get; set; }

    public string Name { get; set; } = string.Empty;
}

public class Person
{
    public int Id { get; set; }

    public string Name { get; set; } = string.Empty;
}
```
Created code
```csharp
public partial class PersonMapper
    {
        [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "4.0.0.0")]
        public partial global::System.Collections.Generic.List<global::AutoInspect.Public.Api.Apis.PublicInspections.ByGuid.Person> ToPerson(global::System.Collections.Generic.List<global::AutoInspect.Public.Api.Apis.PublicInspections.ByGuid.PersonDto> source)
        {
            var target = new global::System.Collections.Generic.List<global::AutoInspect.Public.Api.Apis.PublicInspections.ByGuid.Person>(source.Count);
            foreach (var item in source)
            {
                target.Add(MapToPerson(item));
            }
            return target;
        }

        [global::System.CodeDom.Compiler.GeneratedCode("Riok.Mapperly", "4.0.0.0")]
        private global::AutoInspect.Public.Api.Apis.PublicInspections.ByGuid.Person MapToPerson(global::AutoInspect.Public.Api.Apis.PublicInspections.ByGuid.PersonDto source)
        {
            var target = new global::AutoInspect.Public.Api.Apis.PublicInspections.ByGuid.Person();
            target.Id = source.Id;
            target.Name = source.Name;
            return target;
        }
    }
```

**Reply by reddyabhijith8888** [reply]
@sveinhelge  thank you for the response, i removed the two partial methods and just used these two methods, but my docker build still gives me same errors.
```
public partial List<CustomerItem> MapBLLRequestToDomain (List<Dealer> request); 
public partial List<Dealer> MapDomainRequestToBLL (List<CustomerItem> request);
```

2024-08-01T13:19:14.4979013Z ##[error]D:\a\1\s\myproject\test\test\Mappers\CustomerMapper.cs(12,43): Error CS8795: Partial method 'CustomerMapper.MapBLLRequestToDomain(List<Dealer>)' must have an implementation part because it has accessibility modifiers.
2024-08-01T13:19:14.4990927Z D:\a\1\s\myproject\test\test\Mappers\CustomerMapper.cs(13,68): error CS8795: Partial method 'CustomerMapper.MapDomainRequestToBLL(List<CustomerItem>)' must have an implementation part because it has accessibility modifiers.

**Reply by sveinhelge** [reply]
Some how you get this type of situation.
[Compiler Error CS8795](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs8795)

Have you tried to change the name CustomerMapper to CustomerAbcdefMapper (without using any type of renaming, just change the name directly). So we can rule out other files getting in conflict with this partial class.

**Reply by latonz** [reply]
Having a single and a list mapping for the same types works fine with Mapperly. Mapperly should also be able to identify method signatures with generics correctly, this shouldn't be the issue here...

**Reply by sveinhelge** [reply]
@latonz You are correct. List and single works fine. I think the reason this happend to me was because the compiler was hanging or was delayed. I'm using Rider. After I restarted Rider. This warning was gone after a second or two.
So my mistake suggesting list/single issue.

**Comment by latonz** [comment]
What compiler version are you using in your docker build? To output your compiler version put `#error version` in any of the `.cs` files of the project and run your docker build.

**Reply by cw-joelvarghese** [reply]
I get this error with
CS8304: Compiler version: '4.0.1-1.22181.2 (487283bc)'. Language version: 10.0.

**Comment by ezra-orina-rsa20668** [comment]
Similar issue here. Mine show it has trouble loading Mapperly just before printing the error message in OP. I am on net8, C# v12, and VS 2022.

![image](https://github.com/user-attachments/assets/0637a90e-e008-4610-92fc-039ff79674e6)

**Comment by RytisCepulis** [comment]
Has anybody found the answer to this issue? I am getting the same error

**Comment by fredrikeriksson** [comment]
Not sure this is directly related to OP but maybe the others, i've gotten this issue when analyzers is getting blocked by AppLocker (or similar tooling), generally the builds run fine though.
Analyzers is usually executed in the background in a temp folder like in this case %OSDRIVE%\USERS\*\APPDATA\LOCAL\TEMP\ROSLYN\ANALYZERASSEMBLYLOADER\*\*\RIOK.MAPPERLY.DLL .

If this indeed is the case for you the solution is for your admins to add a AppLocker rule that either whitelist (ordered by the best to least good solution from a security standpoint): 
- The publisher
  *Requires binaries to be code signed, which this library isn't (not the same as strong name signing!!)*
- The path:
  - %OSDRIVE%\USERS\*\APPDATA\LOCAL\TEMP\ROSLYN\ANALYZERASSEMBLYLOADER\*\*\RIOK.MAPPERLY.DLL
   *Do note that any binary with that name at that folder is allowed wether it is the real or not.*
  or
  - %OSDRIVE%\USERS\*\APPDATA\LOCAL\TEMP\ROSLYN\ANALYZERASSEMBLYLOADER\*
    *Should probably be used with caution as this pretty much opens up for any potentially dangerous analyzers.*

Paths might be good enough depending on your threat model and rules, and if you have a good curated internal nuget registry the risk is mitigated somewhat.
